man/zeekctl.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.4.
.TH ZEEKCTL "8" "November 2014" "" "System Administration Utilities"
.SH NAME
.B zeekctl
\- interactive shell for managing Zeek installations
.SH SYNOPSIS
.B zeekctl
[\fIcommand\fR]

.SH DESCRIPTION
.B zeekctl
is an interactive interface for managing either a standalone or a Zeek cluster
installation.  If a \fBzeekctl\fR command is specified directly on the
command-line, then \fBzeekctl\fR performs the action associated with
that command immediately (instead of running the interactive interface).

Before actually running
.B zeekctl
you first need to edit the \fBzeekctl.cfg\fR, \fBnode.cfg\fR, and
\fBnetworks.cfg\fR files. In the \fBzeekctl.cfg\fR file, you should
review the \fBzeekctl\fR options and make sure the options are set
correctly for your environment. Next, edit the \fBnode.cfg\fR
file and specify the nodes that you will be running. Finally, edit
the \fBnetworks.cfg\fR file and list each network that is considered
local to the monitored environment (see the examples
in the file for the format to use).

When running \fBzeekctl\fR for the first time, you must run the \fBzeekctl\fR
\fBdeploy\fR command before running any other commands in order to apply the
configuration settings.  You must also run \fBzeekctl deploy\fR each time
you change the configuration (including any Zeek scripts) or upgrade Zeek.
.SH OPTIONS
.TP
\fBcapstats\fR [<nodes>] [<secs>]
Report interface statistics with capstats
.TP
\fBcheck\fR [<nodes>]
Check configuration before installing it
.TP
\fBcleanup\fR [\-\-all] [<nodes>]
Delete working dirs (flush state) on nodes
.TP
\fBconfig\fR
Print zeekctl configuration
.TP
\fBcron\fR [\-\-no\-watch]
Perform jobs intended to run from cron
.TP
\fBcron\fR enable|disable|?
Enable/disable "cron" jobs
.TP
\fBdeploy\fR
Check, install, and restart
.TP
\fBdf\fR [<nodes>]
Print nodes' current disk usage
.TP
\fBdiag\fR [<nodes>]
Output diagnostics for nodes
.TP
\fBexec\fR <shell cmd>
Execute shell command on all hosts
.TP
\fBexit\fR
Exit from the interactive interface
.TP
\fBinstall\fR
Update zeekctl installation/configuration
.TP
\fBnetstats\fR [<nodes>]
Print nodes' current packet counters
.TP
\fBnodes\fR
Print node configuration
.TP
\fBpeerstatus\fR [<nodes>]
Print status of nodes' remote connections
.TP
\fBprint\fR <id> [<nodes>]
Print values of script variable at nodes
.TP
\fBprocess\fR <trace> [<op>] [\-\- <sc>]
Run Zeek (with options and scripts) on trace
.TP
\fBrestart\fR [\-\-clean] [<nodes>]
Stop and then restart processing
.TP
\fBscripts\fR [\-c] [<nodes>]
List the Zeek scripts the nodes will load
.TP
\fBstart\fR [<nodes>]
Start processing
.TP
\fBstatus\fR [<nodes>]
Summarize node status
.TP
\fBstop\fR [<nodes>]
Stop processing
.TP
\fBtop\fR [<nodes>]
Show Zeek processes ala top
.PP
Commands provided by plugins:
.TP
\fBps.zeek\fR [<nodes>]
Show Zeek processes on nodes' systems
.SH AUTHOR
.B zeekctl
was written by The Zeek Project <info@zeek.org>.


CHANGES

1.9-47 | 2019-05-20 19:37:27 -0700

  * More changes for Bro to Zeek renaming (Daniel Thayer)

1.9-45 | 2019-05-15 15:00:39 -0700

  * Adjust parallelism of build-zeek script (Jon Siwek, Corelight)

  * Update broker.bro module usage to broker.zeek (Jon Siwek, Corelight)

1.9-43 | 2019-05-14 19:29:56 -0700

  * Fix plugin.ps test (Jon Siwek, Corelight)

1.9-42 | 2019-05-14 18:19:43 -0700

  * Remove the "update" command (Jon Siwek, Corelight)

1.9-41 | 2019-05-14 17:27:44 -0700

  * Update Travis config to use zeek/zeekctl (Jon Siwek, Corelight)

  * Update README.rst symlink (Jon Siwek, Corelight)

1.9-39 | 2019-05-14 13:12:15 -0700

  * Fix legacy plugin API and add new tests (Daniel Thayer)

  * Added new test cases and improved a few tests. (Daniel Thayer)

  * Fixed one line in the help output to fit within an 80 character display. (Daniel Thayer)

  * Some fixes for bro-to-zeek renaming and docs (Daniel Thayer)

  * Updating documentation. (Robin Sommer, Corelight)

  * Renamed broctl to zeekctl (Robin Sommer, Corelight)

    I ended up doing the rename pretty comprehensively across all the
    scripts, as it was hard to change some places but not others. So most
    uses of Bro are replaced with Zeek now. I tried to maintain backwards
    compatibility with the old names where user visible, including names
    for options and IDs inside plugins.

    Changes to maintain backwards compabibility:

    - We now also puts links in place for backwards compability:

        bin/broctl -> bin/zeek-wrapper (which then forwards to zeekctl)
        lib/broctl -> lib/zeekctl

    - If an etc/broctl.cfg exists from a previous install, we symlink
      etc/zeekctl.cfg to it to keep any customizations that were made.

    - We create a Python wrapper module BroControl that forwards (with a
      warning) to the renamed ZeekControl, so that old plugins continue to
      work.

    - Old option name containing "Bro" are accepted in place of the new
      Zeek variants.

    - "ps.bro" is an alias for "ps.zeek".

    - BROCTL_DISABLE_LISTEN is an alias for ZEEKCTL_DISABLE_LISTEN

1.9-32 | 2019-05-10 19:13:32 -0700

  * Add LibDir64 option (Jon Siwek, Corelight)

    And make it and LibDir optional dirs for syncing to remote nodes as
    well as for use with the 'df' command.

    On some platforms, certain libraries in the Zeek-ecosystem now install
    into the lib64/ directory by default (per that platform's convention).

  * Allow option names that have numbers in them (Jon Siwek, Corelight)

  * Silence test failures due to rotate_file_by_name deprecation (Jon Siwek, Corelight)

1.9-28 | 2019-04-19 11:11:53 -0700

  * Replace bro_init/bro_done usages with zeek_init/zeek_done (Seth Hall, Corelight)

1.9-24 | 2019-04-16 11:53:06 -0700

  * Update some tests and baselines due to new file extension (Daniel Thayer)

  * Install script files with new file extension (Daniel Thayer)

  * Change file extension of all script files to ".zeek" (Daniel Thayer)

  * Fix the update command

    This broke due to https://github.com/zeek/zeek/pull/261, in which
    errors in initialization are now fatal, but there happened to be
    benign/unnoticed errors with the way `broctl update` was working.
    Namely, it was incorrectly treating the bro process that it spawned
    for using the control framework as a cluster node, by setting the
    CLUSTER_NODE environment variable, and that causes an attempt to listen
    on a port which, when a cluster is up and running as it should be,
    is already listened upon, thus generating an error and completely
    failing now due to change in initialization behavior. (Jon Siwek, Corelight)

1.9-11 | 2019-01-04 13:10:54 -0600

  * Adding support for log rotation/expiration for distributed loggers. (Stefan Maerz)

1.9-8 | 2018-12-10 15:07:31 -0600

  * Change Travis btest commnad to help isolate hung tests (Jon Siwek, Corelight)

1.9-7 | 2018-12-10 13:09:29 -0600

  * Update Travis URL for cloning Zeek (Jon Siwek, Corelight)

  * GH-11: Improve check-pid helper script for Alpine support (Jon Siwek, Corelight)

1.9-4 | 2018-12-07 16:31:33 -0600

  * Update github/download link (Jon Siwek, Corelight)

  * Update submodules to use github.com/zeek (Jon Siwek, Corelight)

1.9-2 | 2018-09-26 10:31:47 -0500

  * Update broctl.rst (by running "make doc") (Daniel Thayer)

1.9 | 2018-09-18 16:47:56 -0500

  * Release 1.9.

1.8-1 | 2018-09-18 16:46:20 -0500

  * Fix commands that use broker python bindings (Jon Siwek, Corelight)

1.8 | 2018-09-18 14:26:09 -0500

  * Release v1.8.

1.7-126 | 2018-09-07 09:56:19 -0500

  * Update the broctl top command to not show the "Proc" column (Daniel Thayer)

  * Update the stats-to-csv script for broker and loggers (Daniel Thayer)

  * Improve `make dist` (Jon Siwek, Corelight)

1.7-122 | 2018-08-29 23:44:14 +0000

  * Include Broker node ID in "control" event topics. This helps break
    a message routing loop due to all "control" nodes being subscribed
    to a common topic. (Jon Siwek, Corelight)

1.7-119 | 2018-08-21 13:20:35 -0500

  * Change default snaplen to 9216 bytes to better accommodate
    jumbo frames (Justin Azoff)

1.7-117 | 2018-08-20 14:57:41 -0500

  * Change broctl to warn about unrecognized broctl options (Daniel Thayer)

  * Remove deprecated node-specific SitePolicy* options (Daniel Thayer)

1.7-107 | 2018-07-16 10:57:58 -0500

  * Update docs about the "update" command being deprecated (Daniel Thayer)

  * Various unit test additions/improvements/cleanups (Daniel Thayer)

  * Update diff-top-output script based on a recent change (Daniel Thayer)

  * Reduce the number of node.cfg files used by the tests (Daniel Thayer)

  * Reduce the number of broctl.cfg files used by the tests (Daniel Thayer)

  * Set BRO_DEFAULT_LISTEN_ADDRESS when running tests (Daniel Thayer)

  * The check and scripts tests no longer need to be serialized (Daniel Thayer)

1.7-93 | 2018-06-21 11:54:09 -0500


ZeekControl/zeekctl.py
# The ZeekControl interactive shell.

from __future__ import print_function
import os
import sys
import logging

from ZeekControl import lock
from ZeekControl import config
from ZeekControl import cmdresult
from ZeekControl import execute
from ZeekControl import control
from ZeekControl import version
from ZeekControl import pluginreg
from ZeekControl import node as node_mod
from ZeekControl.exceptions import *

class TermUI:
    def info(self, txt):
        print(txt)

    def error(self, txt):
        print(txt, file=sys.stderr)
    warn = error


class NullHandler(logging.Handler):
    def emit(self, record):
        pass


def expose(func):
    func.api_exposed = True
    return func

def lock_required(func):
    def wrapper(self, *args, **kwargs):
        self.lock()
        try:
            return func(self, *args, **kwargs)
        finally:
            self.unlock()
    wrapper.lock_required = True
    return wrapper

def lock_required_silent(func):
    def wrapper(self, *args, **kwargs):
        self.lock(showwait=False)
        try:
            return func(self, *args, **kwargs)
        finally:
            self.unlock()
    wrapper.lock_required = True
    return wrapper

def check_config(func):
    def wrapper(self, *args, **kwargs):
        if config.Config.is_cfg_changed():
            self.ui.warn('Configuration has changed. Run the "deploy" command.')
        return func(self, *args, **kwargs)

    return wrapper

class ZeekCtl(object):
    def __init__(self, basedir=version.ZEEKBASE, cfgfile=version.CFGFILE, zeekscriptdir=version.ZEEKSCRIPTDIR, ui=TermUI(), state=None):
        self.ui = ui
        self.zeekbase = basedir

        self.config = config.Configuration(self.zeekbase, cfgfile, zeekscriptdir, self.ui, state)

        # Remove all log handlers (set by any previous calls to logging.*)
        logging.getLogger().handlers = []

        if self.config.debug:
            # Add a log handler that logs to a file.
            try:
                logging.basicConfig(filename=self.config.debuglog,
                            format="%(asctime)s [%(module)s] %(message)s",
                            datefmt=self.config.timefmt,
                            level=logging.DEBUG)
            except IOError as err:
                raise RuntimeEnvironmentError("%s\nCheck if the user running ZeekControl has write access to the debug log file." % err)
        else:
            # Add a log handler that does nothing.
            h = NullHandler()
            logging.getLogger().addHandler(h)

        self.executor = execute.Executor(self.config)
        self.plugins = pluginreg.PluginRegistry()
        self.setup()
        self.controller = control.Controller(self.config, self.ui, self.executor, self.plugins)

    def setup(self):
        plugindirs = self.config.sitepluginpath.split(":")
        plugindirs.append(self.config.plugindir)
        plugindirs.append(self.config.pluginzeekdir)

        for pdir in plugindirs:
            if pdir:
                self.plugins.addDir(pdir)

        self.plugins.loadPlugins(self.ui, self.executor)
        self.plugins.initPluginOptions()
        self.plugins.addNodeKeys()
        self.config.initPostPlugins()
        self.plugins.initPlugins(self.ui)
        self.plugins.initPluginCmds()
        os.chdir(self.config.zeekbase)
        if self.config.get_state("cronenabled") is None:
            self.config.set_state("cronenabled", True)

    def reload_cfg(self):
        self.config.reload_cfg()

        if self.config.debug:
            if isinstance(logging.getLogger().handlers[0], NullHandler):
                # Remove the null handler and configure logging to a file.
                logging.getLogger().handlers = []
                logging.basicConfig(filename=self.config.debuglog,
                            format="%(asctime)s [%(module)s] %(message)s",
                            datefmt=self.config.timefmt,
                            level=logging.DEBUG)

            # Re-enable all log levels.
            logging.disable(logging.NOTSET)
        else:
            # Disable logging to all log levels that we use.
            logging.disable(logging.CRITICAL)

        self.executor.finish()
        self.plugins.initPluginOptions()
        self.config.initPostPlugins()
        self.plugins.initPlugins(self.ui)
        self.plugins.initPluginCmds()

    def finish(self):
        self.executor.finish()
        self.plugins.finishPlugins()

    def warn_zeekctl_install(self):
        self.config.warn_zeekctl_install()

    # Turns node name arguments into a list of nodes.  If "get_hosts" is True,
    # then only one node per host is chosen.  If "get_types" is True, then
    # only one node per node type (manager, proxy, etc.) is chosen.
    def node_args(self, args=None, get_hosts=False, get_types=False):
        nodes = []

        if args:
            for arg in args.split():
                nodelist = self.config.nodes(arg)
                if not nodelist:
                    raise InvalidNodeError("unknown node '%s'" % arg)

                nodes += nodelist

            # Remove duplicate nodes
            newlist = list(set(nodes))
            if len(newlist) != len(nodes):
                nodes = newlist
        else:
            # Get all nodes.
            nodes = self.config.nodes()

        # Sort the list so that it doesn't depend on initial order of arguments
        nodes.sort(key=node_mod.sortnode)

        if get_hosts:
            hosts = {}
            hostnodes = []
            for node in nodes:
                if node.host not in hosts:
                    hosts[node.host] = 1
                    hostnodes.append(node)

            nodes = hostnodes

        if get_types:
            types = {}
            typenodes = []
            for node in nodes:
                if node.type not in types:
                    types[node.type] = 1
                    typenodes.append(node)

            nodes = typenodes

        return nodes

    def lock(self, showwait=True):
        lockstatus = lock.lock(self.ui, showwait)
        if not lockstatus:
            raise LockError("Unable to get lock")

        self.config.read_state()

    def unlock(self):
        lock.unlock(self.ui)

    def node_names(self):
        return [ n.name for n in self.config.nodes() ]

    def node_groups(self):
        return node_mod.node_groups()

    @expose
    @check_config
    def nodes(self):
        results = cmdresult.CmdResult()

        if self.plugins.cmdPre("nodes"):
            for n in self.config.nodes():
                results.set_node_data(n, True, n.to_dict())
        else:
            results.ok = False

        self.plugins.cmdPost("nodes")

        return results

    @expose
    @check_config
    def get_config(self):
        results = cmdresult.CmdResult()

        if self.plugins.cmdPre("config"):
            results.keyval = self.config.options()
        else:
            results.ok = False

        self.plugins.cmdPost("config")
        return results

    @expose
    @check_config
    @lock_required
    def install(self, local=False):
        if self.plugins.cmdPre("install"):
            results = self.controller.install(local)
        else:
            results = cmdresult.CmdResult(ok=False)

        self.plugins.cmdPost("install")
        return results

    @expose
    @check_config
    @lock_required
    def start(self, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("start", nodes)
        results = self.controller.start(nodes)
        self.plugins.cmdPostWithResults("start", results.get_node_data())

        return results

    @expose
    @check_config
    @lock_required
    def stop(self, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("stop", nodes)
        results = self.controller.stop(nodes)
        self.plugins.cmdPostWithResults("stop", results.get_node_data())

        return results

    @expose
    @check_config
    @lock_required
    def restart(self, clean=False, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("restart", nodes, clean)

        self.ui.info("stopping ...")
        results = self.stop(node_list)
        if not results.ok:
            return results

        if clean:
            self.ui.info("cleaning up ...")
            results = self.cleanup(node_list=node_list)
            if not results.ok:
                return results

            self.ui.info("checking configurations ...")
            results = self.check(node_list)
            if not results.ok:
                return results

            self.ui.info("installing ...")
            results = self.install()
            if not results.ok:
                return results

        self.ui.info("starting ...")
        results = self.start(node_list)

        self.plugins.cmdPostWithNodes("restart", nodes)
        return results

    @expose
    @lock_required
    def deploy(self):
        if not self.plugins.cmdPre("deploy"):
            results = cmdresult.CmdResult(ok=False)
            return results

        if self.config.is_cfg_changed():
            self.ui.info("Reloading zeekctl configuration ...")
            self.reload_cfg()

        self.ui.info("checking configurations ...")
        results = self.check(check_node_types=True)
        if not results.ok:
            for (node, success, output) in results.get_node_output():
                if not success:
                    self.ui.info("%s scripts failed." % node)
                    self.ui.info(output)

            return results

        self.ui.info("installing ...")
        results = self.install()
        if not results.ok:
            return results

        self.ui.info("stopping ...")
        results = self.stop()
        if not results.ok:
            return results

        self.ui.info("starting ...")
        results = self.start()

        self.plugins.cmdPost("deploy")
        return results

    @expose
    @check_config
    @lock_required
    def status(self, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("status", nodes)
        results = self.controller.status(nodes)
        self.plugins.cmdPostWithNodes("status", nodes)
        return results

    @expose
    @lock_required
    def top(self, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("top", nodes)
        results = self.controller.top(nodes)
        self.plugins.cmdPostWithNodes("top", nodes)

        return results

    @expose
    @check_config
    @lock_required
    def diag(self, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("diag", nodes)
        results = self.controller.diag(nodes)
        self.plugins.cmdPostWithNodes("diag", nodes)

        return results

    @expose
    @lock_required_silent
    def cron(self, watch=True):
        if self.plugins.cmdPre("cron", "", watch):
            self.controller.cron(watch)
        self.plugins.cmdPost("cron", "", watch)

        return True

    @expose
    @check_config
    @lock_required
    def cronenabled(self):
        results = False
        if self.plugins.cmdPre("cron", "?", False):
            results = self.config.cronenabled
        self.plugins.cmdPost("cron", "?", False)
        return results

    @expose
    @check_config
    @lock_required
    def setcronenabled(self, enable=True):
        if enable:
            if self.plugins.cmdPre("cron", "enable", False):
                self.config.set_state("cronenabled", True)
                self.ui.info("cron enabled")
            self.plugins.cmdPost("cron", "enable", False)
        else:
            if self.plugins.cmdPre("cron", "disable", False):
                self.config.set_state("cronenabled", False)
                self.ui.info("cron disabled")
            self.plugins.cmdPost("cron", "disable", False)

        return True

    @expose
    @check_config
    @lock_required
    def check(self, node_list=None, check_node_types=False):
        nodes = self.node_args(node_list, get_types=check_node_types)

        nodes = self.plugins.cmdPreWithNodes("check", nodes)
        results = self.controller.check(nodes)
        self.plugins.cmdPostWithResults("check", results.get_node_data())

        return results

    @expose
    @check_config
    @lock_required
    def cleanup(self, cleantmp=False, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("cleanup", nodes, cleantmp)
        results = self.controller.cleanup(nodes, cleantmp)
        self.plugins.cmdPostWithNodes("cleanup", nodes, cleantmp)

        return results

    @expose
    @check_config
    @lock_required
    def capstats(self, interval=10, node_list=None):
        nodes = self.node_args(node_list)
        nodes = self.plugins.cmdPreWithNodes("capstats", nodes, interval)
        results = self.controller.capstats(nodes, interval)
        self.plugins.cmdPostWithNodes("capstats", nodes, interval)

        return results

    @expose
    @check_config
    @lock_required
    def df(self, node_list=None):
        nodes = self.node_args(node_list, get_hosts=True)
        nodes = self.plugins.cmdPreWithNodes("df", nodes)
        results = self.controller.df(nodes)
        self.plugins.cmdPostWithNodes("df", nodes)

        return results

    @expose
    @check_config
    @lock_required
    def print_id(self, id, node_list=None):
        nodes = self.node_args(node_list)
        nodes = self.plugins.cmdPreWithNodes("print", nodes, id)
        results = self.controller.print_id(nodes, id)
        self.plugins.cmdPostWithNodes("print", nodes, id)

        return results

    @expose
    @check_config
    @lock_required
    def peerstatus(self, node_list=None):
        nodes = self.node_args(node_list)
        nodes = self.plugins.cmdPreWithNodes("peerstatus", nodes)
        results = self.controller.peerstatus(nodes)
        self.plugins.cmdPostWithNodes("peerstatus", nodes)

        return results

    @expose
    @check_config
    @lock_required
    def netstats(self, node_list=None):
        if not node_list:
            node_list = None
            if not self.config.standalone:
                node_list = node_mod.worker_group()

        nodes = self.node_args(node_list)
        nodes = self.plugins.cmdPreWithNodes("netstats", nodes)
        results = self.controller.netstats(nodes)
        self.plugins.cmdPostWithNodes("netstats", nodes)

        return results

    @expose
    @check_config
    def execute(self, cmd):
        nodes = self.node_args(get_hosts=True)

        if self.plugins.cmdPre("exec", cmd):
            results = self.controller.execute_cmd(nodes, cmd)
        else:
            results = cmdresult.CmdResult(ok=False)

        self.plugins.cmdPost("exec", cmd)

        return results

    @expose
    @check_config
    @lock_required
    def scripts(self, check=False, node_list=None):
        nodes = self.node_args(node_list)

        nodes = self.plugins.cmdPreWithNodes("scripts", nodes, check)
        results = self.controller.scripts(nodes, check)
        self.plugins.cmdPostWithNodes("scripts", nodes, check)

        return results

    @expose
    @check_config
    @lock_required
    def process(self, trace, options, scripts):
        if self.plugins.cmdPre("process", trace, options, scripts):
            results = self.controller.process(trace, options, scripts)
        else:
            results = cmdresult.CmdResult(ok=False)

        self.plugins.cmdPost("process", trace, options, scripts, results.ok)

        return results

    @expose
    @check_config
    @lock_required
    def plugincmd(self, cmd, args):
        return self.plugins.runCustomCommand(cmd, args, self.ui)

